library-backend/library-app/src/main/java/org/internship/library/app/security/MyUserDetailsService.java

```diff
@@ -0,0 +1,97 @@
+    UserRepository userRepository;
+    @Autowired
+    ApplicationPasswordEncoder applicationPasswordEncoder;
+
+    private final String PASSWORD_PATTERN =
+            "^(?=.*[0-9])(?=.*[a-z])(?=.*[A-Z])(?=.*[!@#&()–[{}]:;',?/*~$^+=<>]).{4,20}$";
+    private final String INVALID_CHARACTERS = "^[^<>'\"/;:+=&`%{}\\]\\[\\\\?]*$";
+
+    private final Pattern invalidPattern = Pattern.compile(INVALID_CHARACTERS);
+    private final Pattern pattern = Pattern.compile(PASSWORD_PATTERN);
+
+    @Override
+    public UserDetails loadUserByUsername(String userName) throws UsernameNotFoundException {
+           Optional<UserEntity> user = userRepository.findByUserName(userName);
+
+           user.orElseThrow(() -> new UsernameNotFoundException("Not found: " + userName));
+
+//           return user.map(MyUserDetails::new).get();
+            return new MyUserDetails(user.get());
+    }
+
+    public UserEntity signUp(UserEntity user) {
+        logger.info("Sign Up the New User with the username: " + user.getUserName());
+
+        Optional<UserEntity> userExists = userRepository.findByUserName(user.getUserName());
+
+        if(userExists.isPresent()) {
+            throw new UserCredentialsException("USERNAME EXISTS");
+        }
+
+        Optional<UserEntity> emailExists = userRepository.findByEmail(user.getEmail());
+
+        if(emailExists.isPresent()) {
+            throw new UserCredentialsException("EMAIL EXISTS");
+        }
+
+//        validatePassword(user);
+        String encodedPassword = applicationPasswordEncoder.encode(user.getPassword());
+        user.setPassword(encodedPassword);
+        userRepository.save(user);
+
+        return user;
+    }
+
+    private boolean isValid(String password) {
+        Matcher matcher = pattern.matcher(password);
+        return matcher.matches();
+    }
+
+    private boolean invalidCharacters(String password) {
+        Matcher matcher = invalidPattern.matcher(password);
+        return matcher.matches();
+    }
+
+    private void validatePassword(UserEntity user) {
+
+        if (user.getPassword().trim().isEmpty()) {
+            throw new UserCredentialsException("PASSWORD IS EMPTY");
+        }
+
+        if (!invalidCharacters(user.getPassword())) {
+            throw new UserCredentialsException(
+                    "CONTAINS_INVALID_CHARACTER [';', '<', '>', '{', '}', '[', ']', '+', '=', '?', '&', ':', '\\','`']");
+        }
+
+        if (!isValid(user.getPassword())) {
+            throw new UserCredentialsException("Password must contain at least one digit [0-9]." +
+                    "Password must contain at least one lowercase Latin character [a-z]." +
+                    "Password must contain at least one uppercase Latin character [A-Z]." +
+                    "Password must contain at least one special character like ! @ # & ( )." +
+                    "Password must contain a length of at least 8 characters and a maximum of 20 characters.");
+        }
+    }
```

Comment:
The password validation implementation could be done outside this service. One option would be to move the code to a _PasswordValidationHelper_ class and here you just simply call a static method _PasswordValidationHelper.validate(password)_.
Or you could reimplement the validation using _Passay_ library, take a look [here](https://www.baeldung.com/registration-password-strength-and-rules).

Reply:
I implemented Passay, but it doesn't work because if I put @ValidPassword on password in UserEntity, the rules don't apply because it verifies the encrypted password by BCrypt, so every password will be ok